### DISCUSSION TITLE
Firestore 텍스트 검색 이슈

### DISCUSSION THREAD

**Discussion by Jeong-heon2** [boostcampwm-2021/Android08-Ratatouille#47] [Issue Talk] [answered]
## 이슈 
https://firebase.google.com/docs/firestore/solutions/search?hl=ko
Cloud Firestore는 기본 색인 생성 또는 문서의 텍스트 필드 검색을 지원하지 않습니다. ( Like 쿼리 지원 안 됨 ) 
또한 클라이언트 측에서 필드를 검색하고자 전체 컬렉션을 다운로드하는 방법은 실용적이지 않습니다.
Cloud Firestore 데이터의 전체 텍스트 검색을 지원하려면 타사 전용 검색 서비스를 사용하라고 나와있습니다. 

## 대안 

1. 아래와 비슷한 방법으로 prefix 만 검색하는 방법이 있는데 한글에서도 가능할지 의문입니다. 
```kotlin
db.collection(c)
    .where('foo', '>=', 'bar')
    .where('foo', '<', 'bas');
   // bar  ~ bas 까지 다 가져온다 ( ex. bara, barb, barab ...  ) 
```

2. 검색 기능을 빼고 필터링 기능 강화 
기존의 최신순, 인기순 필터링 +  레시피 종류 필터링 기능 ( 한식, 중식 .. ) 

이 외에도 더 좋은 방법이 있을까요?

**Comment by svclaw2000** [comment]
```kotlin
ref.where('title', '>=', term).where('title', '<=', term + '~');
```
이렇게 하면 한글 영어 상관없지 않을까요??
term으로 시작하고, 뒤에 뭐가 오든 상관없도록 하는 query인 것 같습니다!

**Reply by Jeong-heon2** [reply]
where 이라는 함수가 없어요! 
```kotlin
firestore.collection(RECIPE_COLLECTION_ID)
                .whereGreaterThanOrEqualTo(FIELD_TITLE, keyword)
                .whereLessThan(FIELD_TITLE, keyword + "~")
                .orderBy(FIELD_TITLE, Query.Direction.ASCENDING)
                .startAfter(lastVisibleTitle)
                .limit(PAGING_SIZE)
```
이런식으로 해야 할 것 같습니다 , 
그런데 위와 같이 하니까 동작 안 하네요,  "~" 을  "힣"으로 바꾸니까 되긴 합니다

**Reply by svclaw2000** [reply]
그렇군요..! 제가 본 글이 영어에 대한거였는데, 한글이 적용되지 않나보네요...

**Comment by Jeong-heon2** [comment]
대안 1 확인 결과 한글로도 가능하긴 합니다
검색 키워드 '김치' 일때  김치 <=     < 김치힣   인  것을 가져오도록 하면 됩니다,

**Comment by Jeong-heon2** [comment, marked as answer]
```kotlin
firestore.collection(RECIPE_COLLECTION_ID)
          .whereGreaterThanOrEqualTo(FIELD_TITLE, keyword)
          .whereLessThan(FIELD_TITLE, keyword + "힣")
          .orderBy(FIELD_TITLE, Query.Direction.ASCENDING)
          .startAfter(lastVisibleTitle)
          .limit(PAGING_SIZE)
          .get()
          .await()
          .map { queryDocumentSnapshot ->
              entityToDomain(queryDocumentSnapshot.toObject<RecipeEntity>())
          }
```
### DB  ( title) 
"맛있는 김치a"
"맛있는 김치b"
"맛있는 김치찌개" 
### keyword
맛있는
### lastVisibleTitle
맛있는 김치a
### 결과 
"맛있는 김치b"
"맛있는 김치찌개" 


맛있는 ~ 맛있는힣  에 매칭되는 title를 사전 오름차순으로 정렬 후, 
lastVisibleTitle 이후 PAGING_SIZE 만큼 가져옵니다.  lastVisibleTitle 이 빈 문자열 "" 일때, 정렬 한 거에서 처음꺼 부터 가져옵니다.
